src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -212,22 +212,38 @@ private int GetSegmentSize(int sizeHint)
 
         private BufferSegment CreateSegmentUnsynchronized()
         {
-            if (_pooledSegmentCount > 0)
+            int index = _pooledSegmentCount - 1;
+            BufferSegment[] pool = _bufferSegmentPool;
+            if ((uint)index < (uint)pool.Length)
             {
-                _pooledSegmentCount--;
-                return _bufferSegmentPool[_pooledSegmentCount];
+                _pooledSegmentCount = index;
+                return pool[index];
             }
 
             return new BufferSegment();
         }
 
-        private void ReturnSegmentUnsynchronized(BufferSegment segment)
+        private void ReturnSegmentsUnsynchronized(BufferSegment current, BufferSegment endExclusive)
         {
-            if (_pooledSegmentCount < _bufferSegmentPool.Length)
+            int count = _pooledSegmentCount;
+            BufferSegment[] pool = _bufferSegmentPool;
+
+            while (current != null && current != endExclusive)
             {
-                _bufferSegmentPool[_pooledSegmentCount] = segment;
-                _pooledSegmentCount++;
+                BufferSegment next = current.NextSegment;
+                current.ResetMemory();
+
+                if ((uint)count < (uint)pool.Length)
```

Comment:
Thought about breaking out of the loop here and entering an new one without this check; but at that point you have 16 segments in the pool and still have more to return, so should be on an uncommon path; to introduce a second loop that is immediately checked and skipped to the common path.